### ISSUE TITLE
UI Relative Transformations

### ISSUE THREAD

**Issue by gmarais** [godotengine/godot-proposals#413]
### **Describe the project you are working on:**

I am making a simple mobile game where half of my UI is placed at the bottom right of the screen and mimics in the way it is disposed the one at the top but instead of growing from the top left it grows from the bottom right.

Think of a top down card game where both players have their decks on their right and the cards played are positioned in front of them from their right to their left on two or more rows and all cards are facing the bottom player so that he can read and look at them.

The reference points change but the way the cards should be placed from each reference point is pretty much the same.

### **Describe the problem or limitation you are having in your project:**

As it stands right now in godot, UI nodes are positioned relative to the top left corner of their parent position meaning (0,0) is the top left corner of the ui node.

While this is not a problem when you place your elements from top to bottom and left to right this process becomes much more troublesome, when you want to do the same from the bottom-up and/or from the right to the left especially if you want your nodes to freely grow and shrink or resize in these directions.

On top of that many problems also arise with nodes like gridcontainers that always fill left to right and then top to bottom.
If you place the same node on your bottom right corner and want it to grow from bottom to top and right to left you are in trouble.
HBox also has a weird behavior although this node does allow you to specify a direction to fill, childs are sorted backward meaning first child will be on the left even if it fills from the right etc...
Also the grow direction  only applies to the min size not the size of the object.

Say I dont know how many elements will be in that UI node I just placed on the bottom right of my screen and/or I dont know the height or the width it will have after more elements are added to my containers since it is supposed to behave and grow in size with its content this becomes even more problematic.

Also in order to precisely place my objects in such a UI node I have to always place them from the top left position of the parent in the inspector I could place them with the margins but since I cant lock the size parameter if I change the margins to place them using that the size will be changed and not the position.

Overall a lot of things are there but they are not working as well as maybe they could be relative to one another allowing for more flexibility while being more intuitive when the user decides to change his reference point.

### **Describe how this feature / enhancement will help you overcome this problem or limitation:**

Here is what I suggest which I think would make it work more evenly no matter what orientation and place you want to build your ui from in your screen:

In the inspector:
The position of the ui node rectangle to become relative to it's pivot point.
The pivot point to be defined relatively to the ui node boundaries (0.5, 0.5) being the center.
A pivot defined as (0, 0) could be the top left corner of the node is the pivot point, and (1, 1) would be the bottom right corner.
The position of the pivot to be relative to the ui node anchors to keep all the responsiveness.
The growing direction would then not need to be set since it would just be defined by the position of the pivot in relation to the ui node rectangle if my pivot is at the center it grows in all directions if it is off centered it grows so that the pivot keeps its position within my ui node.

The pivot therefore determines how size and scale affect the object making it very easy to have an object scale up and down in an animation for example while staying properly centered or off centered.

The position of my ui node would then just reflect the distance of my pivot from the anchor's center point.
(if my left anchor is 0.5 and my right is 1 then my center point is at 0.75 of my width etc)

This would allow my objects to be placed from the bottom right exactly the same way we do from the top left.

That way at (0,0) my object is on the bottom right and I can start positioning my objects relatively to that corner instead which will be fixed unlike the top left corner that may vary with the size of my objects.

The idea here is that the anchors are the only variable linked to my parent node all is relative to them from there making it very easy to move the ui around in a responsive and scalable manner it just is a manner of changing the pivots and anchors and all works the same again.

This feature would also make very handy the possibility to lock properties to always know when the user makes a change one another property what will be affected and changed along and what will be kept.

### **Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

This image shows how each variable represents the position of the node content relative to its pivot and anchors in a more intuitive and scallable way since what will work for aligning your objects from the top left should work exactly the same from the bottom right without having to script your way through the trouble.

<img src="https://svgshare.com/i/HbG.svg" />

### **Describe implementation detail for your proposal (in code), if possible:**

Add relative variables to the ui node and make anychange to these relative variables try to affect the position rotation size and scale of the object while keeping the locked parameters values if too many parameters are locked for the change to be possible just ignore the change.
Expose these relative parameters to the inspector while

### **If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Making these variables relative to each other and react in such a way in the editor would require both a plugin for the inspector and a plugin for each nodes just to make something more flexible and more intuitive.

### **Is there a reason why this should be core and not an add-on in the asset library?:**
This feature's main goal is to make the engine more flexible and intuitive, while it should not change drastically the core underlying mechanics it does at least change the way we interact with these variables from the editor but also from scripts in a very systematic manner.

**Comment by groud** [comment]
Your initial problem of growing UI is soved by the grow_direction property of the HBoxContainer and VBoxContainer nodes. Have you tried using them ?

**Comment by gmarais** [comment]
Yes I did but as I just described even then there are issues with Hbox and Vbox like:

The Alignement begin or end works for these elements but say I have 3 nodes in my containers named 1 2 3 if I use the grow direction begin they sill be moved on the right as expected but they will still be sorted from left to right which is not consistent.

Basically with a little picture:
|123    |
becomes
|    123|
instead of
|    321|

The grow direction however only applies to the changes done to the min_size but any changes you do to the size or scale are not affected by that variable.
Also it cannot handle more cases like if you want your element to grow 2 times more on the right than on the left or other trivial cases like that.

Grow direction also does not affect the reference point from which child objects are placed in it, it stays at the top left of the parent object.

If that makes more sense. Thank you for trying to help :-)
